sql/expression/function/trim_ltrim_rtrim.go

```diff
@@ -18,91 +18,167 @@ import (
 	"fmt"
 	"reflect"
 	"strings"
-	"unicode"
 
 	"github.com/dolthub/go-mysql-server/sql"
 	"github.com/dolthub/go-mysql-server/sql/expression"
 )
 
-type trimType rune
+type Trim struct {
+	str sql.Expression
+	pat sql.Expression
+	dir sql.Expression
+}
 
-const (
-	lTrimType trimType = 'l'
-	rTrimType trimType = 'r'
-	bTrimType trimType = 'b'
-)
+var _ sql.FunctionExpression = (*Trim)(nil)
 
-func NewLeftTrim(e sql.Expression) sql.Expression {
-	return newTrim(lTrimType, e)
+func NewTrim(str, pat, dir sql.Expression) sql.Expression {
+	return &Trim{str, pat, dir}
 }
 
-func NewRightTrim(e sql.Expression) sql.Expression {
-	return newTrim(rTrimType, e)
+// FunctionName implements sql.FunctionExpression
+func (t *Trim) FunctionName() string {
+	return "trim"
 }
 
-func NewTrim(e sql.Expression) sql.Expression {
-	return newTrim(bTrimType, e)
+// Children implements the Expression interface.
+func (t *Trim) Children() []sql.Expression {
+	return []sql.Expression{t.str, t.pat, t.dir}
 }
 
-// newTrim creates a new Trim expression.
-func newTrim(tType trimType, str sql.Expression) sql.Expression {
-	return &Trim{expression.UnaryExpression{Child: str}, tType}
-}
+// Eval implements the Expression interface.
+func (t *Trim) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	// Evaluate pattern
+	pat, err := t.pat.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
 
-// Trim is a function that returns the string with prefix or suffix spaces removed based on the trimType
-type Trim struct {
-	expression.UnaryExpression
-	trimType
-}
+	// Cast to string
+	var pat_text string
```

Comment:
Some more tests are in order, but you probably want to use Convert() of a string type rather than a simple type assertion like this

E.g. this works:

```
mysql> SELECT TRIM(LEADING 1 FROM '111barxxx');
+----------------------------------+
| TRIM(LEADING 1 FROM '111barxxx') |
+----------------------------------+
| barxxx                           |
+----------------------------------+
```